core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -317,6 +377,17 @@ final class ShuffleBlockFetcherIterator(
         hostLocalBlockBytes += mergedBlockInfos.map(_.size).sum
       } else {
         remoteBlockBytes += blockInfos.map(_._2).sum
+        val inMemoryBlocks = blockInfos.filter(_._2 <= maxReqSizeShuffleToMem).map(_._2)
+        val maxInMemoryBlock = if (inMemoryBlocks.isEmpty) 0L else inMemoryBlocks.max
+        // scalastyle:off
+        println(PlatformDependent.maxDirectMemory())
+        println(maxInMemoryBlock)
+        if (PlatformDependent.directBufferPreferred() &&
+            PlatformDependent.maxDirectMemory() < maxInMemoryBlock) {
+          throw new SparkException(
```

Comment:
This is a very corner case. Probably we can just check if the netty memory is larger than `maxReqSizeShuffleToMem`